zerver/lib/upload.py

```diff
@@ -425,6 +428,12 @@ def delete_file_from_s3(self, path_id: str, bucket: ServiceResource) -> bool:
         key.delete()
         return True
 
+    def get_public_upload_root_url(self) -> str:
+        # boto requires a key name, so we can't just pass "" here;
+        # trim the offending character back off from the URL we get
+        # back.
+        return self.get_public_upload_url("a")[:-1]
```

Comment:
I'm not happy with this, but I'm short on other ideas.

Reply:
I didn’t find another idea. But given how much of a black box Boto3 is, the fact that it sometimes returns a query string, and the potential for variation among non-Amazon S3-compatible providers, I feel like we should attempt some better parsing and sanity checking here.

```python
u = urlsplit(self.get_public_upload_url("a"))
assert u.path.endswith("/a")
return urlunsplit((u.scheme, u.netloc, u.path[:-1], "", ""))
```

Reply:
Yeah, makes sense to me.